src/services/bunker.py

```diff
@@ -1,15 +1,344 @@
+            self.w3.lido_contracts.withdrawal_vault.address, vault_balance_at.block_hash
+        )
+
+        return lido_balance, self.w3.from_wei(withdrawal_vault_balance, 'gwei')
+
+    def _get_withdrawn_from_vault_between(self, prev_blockstamp: BlockStamp, curr_blockstamp: BlockStamp) -> int:
+        """
+        Lookup for ETHDistributed event and sum up all withdrawalsWithdrawn
+        """
+        prev_epoch = prev_blockstamp.slot_number // self.c_conf.slots_per_epoch
+        curr_epoch = curr_blockstamp.slot_number // self.c_conf.slots_per_epoch
+        logger.info({"msg": f"Get withdrawn from vault between {prev_epoch,curr_epoch=}"})
+        vault_withdrawals = 0
+        events = self.w3.lido_contracts.lido.events.ETHDistributed.get_logs(
+            from_block=int(prev_blockstamp.block_number) + 1, to_block=int(curr_blockstamp.block_number)
+        )
+        for event in events:
+            vault_withdrawals += event['args']['withdrawalsWithdrawn']
+
+        vault_withdrawals = self.w3.from_wei(vault_withdrawals, 'gwei')
+
+        logger.info({"msg": f"Vault withdrawals: {vault_withdrawals} Gwei"})
+
+        return vault_withdrawals
+
+    @staticmethod
+    def _calculate_total_effective_balance(all_validators: list[Validator]) -> Gwei:
+        """
+        Calculates total balance of all active validators in network
+        """
+        total_effective_balance = 0
+
+        for v in all_validators:
+            if v.status in [
+                ValidatorStatus.ACTIVE_ONGOING,
+                ValidatorStatus.ACTIVE_EXITING,
+                ValidatorStatus.ACTIVE_SLASHED
+            ]:
+                total_effective_balance += int(v.validator.effective_balance)
+
+        return total_effective_balance
+
+    @staticmethod
+    def _not_withdrawn_slashed_validators(all_validators: list[Validator], ref_epoch: Epoch) -> list[Validator]:
+        """
+        Get all slashed validators, who are not withdrawn yet
+        """
+        slashed_validators = []
+
+        for validator in all_validators:
+            v = validator.validator
+            if v.slashed and int(v.withdrawable_epoch) > ref_epoch:
+                slashed_validators.append(validator)
+
+        return slashed_validators
+
+    @staticmethod
+    def _detect_slashing_epoch_range(validator: Validator, ref_epoch: Epoch) -> Sequence[Epoch]:
+        """
+        Detect slashing epoch range for validator
+        If difference between validator's withdrawable epoch and exit epoch is greater enough,
+        then we can be sure that validator was slashed in particular epoch
+        Otherwise, we can only assume that validator was slashed in epochs range
+        due because its exit epoch shifted because of huge exit queue
+        Read more here: https://hackmd.io/@lido/r1Qkkiv3j
+        """
+        assert validator.validator.slashed, "Validator should be slashed to detect slashing epoch range"
```

Comment:
No asserts. Raise exceptiopns
Asserts could be turned off in production code.

Reply:
done